sw-emulator/lib/periph/src/ml_dsa87.rs

```diff
@@ -0,0 +1,1254 @@
+/*++
+
+Licensed under the Apache-2.0 license.
+
+File Name:
+
+ml_dsa87.rs
+
+Abstract:
+
+File contains Ml_Dsa87 peripheral implementation.
+
+--*/
+
+use caliptra_emu_bus::{ActionHandle, BusError, Clock, ReadOnlyRegister, ReadWriteRegister, Timer};
+use caliptra_emu_derive::Bus;
+use caliptra_emu_types::{RvData, RvSize};
+use fips204::ml_dsa_87::{try_keygen_with_rng, PrivateKey, PublicKey, PK_LEN, SIG_LEN, SK_LEN};
+use fips204::traits::{SerDes, Signer, Verifier};
+use rand::rngs::StdRng;
+use rand::SeedableRng;
+use tock_registers::interfaces::{ReadWriteable, Readable, Writeable};
+use tock_registers::register_bitfields;
+
+use crate::helpers::{bytes_from_words_le, words_from_bytes_le};
+
+/// ML_DSA87 Initialization Vector size
+const ML_DSA87_IV_SIZE: usize = 64;
+
+/// ML_DSA87 Key Generation seed
+const ML_DSA87_SEED_SIZE: usize = 32;
+
+/// ML_DSA87 SIGN_RND size
+const ML_DSA87_SIGN_RND_SIZE: usize = 32;
+
+/// ML_DSA87 MSG size
+const ML_DSA87_MSG_SIZE: usize = 64;
+
+/// ML_DSA87 VERIFICATION size
+const ML_DSA87_VERIFICATION_SIZE: usize = 64;
+
+/// The number of CPU clock cycles it takes to perform Ml_Dsa87 operation
+const ML_DSA87_OP_TICKS: u64 = 1000;
+
+register_bitfields! [
+    u32,
+
+    /// Control Register Fields
+    Control [
+        CTRL OFFSET(0) NUMBITS(2) [
+            IDLE = 0b00,
+            GEN_KEY = 0b01,
+            SIGN = 0b10,
+            VERIFY = 0b11,
+        ],
+        ZEROIZE OFFSET(2) NUMBITS(1) [],
+//        PCR_SIGN OFFSET(3) NUMBITS(1) [] TODO remove right??
+    ],
+
+    /// Status Register Fields
+    Status[
+        READY OFFSET(0) NUMBITS(1) [],
+        VALID OFFSET(1) NUMBITS(1) [],
+        RSVD OFFSET(2) NUMBITS(30) [],
+    ],
+];
+
+#[derive(Bus)]
+#[poll_fn(poll)]
+#[warm_reset_fn(warm_reset)]
+#[update_reset_fn(update_reset)]
+pub struct MlDsa87 {
+    /// Name 0 register
+    #[register(offset = 0x0000_0000)]
+    name0: ReadOnlyRegister<u32>,
+
+    /// Name 1 register
+    #[register(offset = 0x0000_0004)]
+    name1: ReadOnlyRegister<u32>,
+
+    /// Version 0 register
+    #[register(offset = 0x0000_0008)]
+    version0: ReadOnlyRegister<u32>,
+
+    /// Version 1 register
+    #[register(offset = 0x0000_000C)]
+    version1: ReadOnlyRegister<u32>,
```

Comment:
nit: For these multi-word registers, is it easier to just model it as an array of registers?

Reply:
DOne